### PR TITLE
Mesh 1865/asset metadata improvements

### PR DESCRIPTION
**Pull request by HenriqueNogara**
Adds two extrinsics: `remove_local_metadata_key`and `remove_metadata_value`.

### new features

-  `remove_local_metadata_key` extrinsic;
-  `remove_metadata_value` extrinsic;

### modified API

- Adds two error variants: `AssetMetadataKeyBelongsToNFTCollection`, `AssetMetadataValueIsEmpty`
- Adds two events: `MetadataValueDeleted`, `LocalMetadatakeyDeleted`

### modified logic

- `base_set_asset_metadata_details` no longer allows locking a key with empty value.

### CODE REVIEW

File: pallets/asset/src/lib.rs [hunk 1]
```diff
@@ -911,6 +912,47 @@ decl_module! {
...
+        ///
+        /// # Permissions
+        /// * Asset
+        #[weight = <T as Config>::WeightInfo::remove_local_metadata_key()]
+        pub fn remove_local_metadata_key(origin, ticker: Ticker, name: AssetMetadataName) -> DispatchResult {
+            Self::base_remove_local_metadata_key(origin, ticker, name)
+        }
+
+        /// Removes the asset metadata value of a local key.
+        ///
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: The metadata value can be removed for either local or global keys.

We just can't remove the definition of global keys.

File: pallets/asset/src/lib.rs [hunk 2]
```diff
@@ -2190,6 +2236,14 @@ impl<T: Config> Module<T> {
             Error::<T>::AssetMetadataValueIsLocked
         );
 
+        // Prevent locking an asset metadata with no value
+        if let AssetMetadataLockStatus::Locked | AssetMetadataLockStatus::LockedUntil(_) =
+            detail.lock_status
+        {
+            AssetMetadataValues::try_get(&ticker, &key)
+                .map_err(|_| Error::<T>::AssetMetadataValueIsEmpty)?;
+        }
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: I think a `match` with all enum variants would be safer here.  If we add a new "Locked" variant in the future, the compiler can help catch it.  Just don't use the catch-all `_ => (),` pattern.

Or just call `detail.lock_status.is_locked(0)`.

File: pallets/common/src/traits/asset.rs [hunk 3]
```diff
@@ -242,5 +248,11 @@ decl_event! {
         /// An event emitted when the type of an asset changed.
         /// Parameters: caller DID, ticker, new token type.
         AssetTypeChanged(IdentityId, Ticker, AssetType),
+        /// An event emitted when a local metadata key has been removed.
+        /// Parameters: caller ticker, Local type name
+        LocalMetadataKeyDeleted(Ticker, AssetMetadataName),
+        /// An event emitted when a local metadata value has been removed.
+        /// Parameters: caller ticker, Local type name
+        LocalMetadataValueDeleted(Ticker, AssetMetadataName),
```
Comment (Neopallium) [inline pallets/common/src/traits/asset.rs]: These should include the caller's DID as the first parameter.

File: pallets/asset/src/lib.rs [hunk 4]
```diff
@@ -911,6 +912,47 @@ decl_module! {
         pub fn update_asset_type(origin, ticker: Ticker, asset_type: AssetType) -> DispatchResult {
             Self::base_update_asset_type(origin, ticker, asset_type)
         }
+
+        /// Removes a local metadata key.
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]:
```suggestion
        /// Removes a local metadata key and value.
```

File: pallets/asset/src/lib.rs [hunk 5]
```diff
@@ -911,6 +912,47 @@ decl_module! {
...
+        ///  - `SecondaryKeyNotAuthorizedForAsset` - if called by someone without the appropriate external agent permissions.
+        ///  - `UnauthorizedAgent` - if called by someone without the appropriate external agent permissions.
+        ///  - `AssetMetadataKeyIsMissing` - if the key doens't exist.
+        ///  - `AssetMetadataValueIsLocked` - if the value of the key is locked.
+        ///  - AssetMetadataKeyBelongsToNFTCollection - if the key is a mandatory key in an NFT collection.
+        ///
+        /// # Permissions
+        /// * Asset
+        #[weight = <T as Config>::WeightInfo::remove_local_metadata_key()]
+        pub fn remove_local_metadata_key(origin, ticker: Ticker, name: AssetMetadataName) -> DispatchResult {
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: Should use `AssetMetadataLocalKey` instead of the name.

**Review by Neopallium — CHANGES_REQUESTED** [review]

**Review by adamdossa — APPROVED** [review]

File: pallets/asset/src/lib.rs [hunk 6]
```diff
@@ -2596,4 +2648,75 @@ impl<T: Config> Module<T> {
...
+        // Verifies if the caller has the correct permissions for this asset
+        let caller_did = <ExternalAgents<T>>::ensure_perms(origin, ticker)?;
+        // Verifies if the key exists.
+        match metadata_key {
+            AssetMetadataKey::Global(global_key) => {
+                let _name = AssetMetadataGlobalKeyToName::try_get(&global_key)
+                    .map_err(|_| Error::<T>::AssetMetadataKeyIsMissing)?;
+            }
+            AssetMetadataKey::Local(local_key) => {
+                let _name = AssetMetadataLocalKeyToName::try_get(ticker, &local_key)
```
Comment (Neopallium) [inline pallets/asset/src/lib.rs]: Please use `contains_key` here.

**Review by Neopallium — APPROVED** [review]

**Pull request merged**